_posts/2020-02-25-password-encoding-spring-security.md

```diff
@@ -0,0 +1,535 @@
+password to the encoded form and the method `matches()` to compare the plain password with the encoded password.
+Every encoder has a default constructor,
+that creates an instance with the default work factor. We can use other constructors for tuning the work factor.
+
+
+### BCryptPasswordEncoder
+```java
+ int strength = 10; // work factor of bcrypt
+ BCryptPasswordEncoder bCryptPasswordEncoder =
+    new BCryptPasswordEncoder(strength, new SecureRandom());
+ String encodedPassword = bCryptPasswordEncoder.encode(plainPassword);
+```
+
+`BCryptPasswordEncoder` has the parameter `strength`. The default value in Spring Security is 10.
+It is recommended to use the `SecureRandom`, because it provides a cryptographically strong random number for a salt generation.
+The output looks like
+```
+$2a$10$EzbrJCN8wj8M8B5aQiRmiuWqVvnxna73Ccvm38aoneiJb88kkwlH2
+```
+
+Note, in contrast to simple hash algorithms like SHA-256 or MD5, the output of `bcrypt` contains meta information
+about the version of the algorithm, work factor, and salt. We don't need to save this information separately.
+
+### Pbkdf2PasswordEncoder
+````java
+String pepper = "pepper"; // secret key used by password encoding
+int iterations = 200000; // number of hash iteration
+int hashWidth = 256; // hash width in bits
+
+Pbkdf2PasswordEncoder pbkdf2PasswordEncoder =
+        new Pbkdf2PasswordEncoder(pepper, iterations, hashWidth);
+pbkdf2PasswordEncoder.setEncodeHashAsBase64(true);
+String encodedPassword = pbkdf2PasswordEncoder.encode(plainPassword);
+````  
+
+The PBKDF2 algorithm was not designed for password encoding but the key derivation from a password.
+It runs the hash algorithm over the plain password many times. It generates a salt too. We
+can define how long the output can be and additionally we can use a secret called `pepper` to
+make the password encoding more secure.
+
+The output looks like
+
+```
+lLDINGz0YLUUFQuuj5ChAsq0GNM9yHeUAJiL2Be7WUh43Xo3gmXNaw==
+```
+
+The salt is saved within, but we have to save the iteration number and hash width additionally. The `pepper`
+should be kept as a secret.
+The default value of the iteration number is 185000 and the hash width is 256
+
+### SCryptPasswordEncoder
+```java
+int cpuCost = (int) Math.pow(2, 14); // factor to increase CPU costs
+int memoryCost = 8; // factor to increases memory usage
+int parallelization = 1; // currently  nor supported by Spring Security
+int keyLength = 32; // key length in bytes
+int saltLength = 64; // salt length in bytes
+
+SCryptPasswordEncoder sCryptPasswordEncoder =
+        new SCryptPasswordEncoder(cpuCost, memoryCost, parallelization, keyLength, saltLength);
+String encodedPassword = sCryptPasswordEncoder.encode(plainPassword);
+```
+
+The `scrypt` algorithm can not only configure the CPU cost but memory cost. It makes the attack very expensive.
+For example with the parameter from the code above, the password encoding will require 16MB memory.   
```

Comment:
This is not the memory size, it is a factor to increase memory usage. The algorithm is pretty complicated, I didn't want to explain it here. I added a short explanation.

Reply:
> In this encoder the memory and CPU cost depends on each other and can not be tuned separately

But we DID configure them separately using `cpuCost` and `memoryCost` parameters. I still don't understand this sentence.

Reply:
I get it now...the `memoryCost` parameter is just a factor and doesn't specify the max memory. I removed the sentence alltogether because if you talk about 16MB here, the reader makes a connection to the `memoryCost` parameter like I did.